stl/inc/xutility

```diff
@@ -4124,12 +4124,51 @@ _CONSTEXPR20 _OutIt _Copy_unchecked(_InIt _First, _InIt _Last, _OutIt _Dest) {
     return _Dest;
 }
 
+// _Iterators_are_contiguous<_Iter1, _Iter2> reports whether both iterators are known to be contiguous.
+// (Without concepts, this detection is limited, which will limit when we can activate the memcmp optimization.)
+
+#ifdef __cpp_lib_concepts
+// When concepts are available, we can detect arbitrary contiguous iterators.
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = contiguous_iterator<_Iter1> //
+    && contiguous_iterator<_Iter2>;
+
+template <class _Target, class _Iter>
+_NODISCARD constexpr _Target* _To_pointer(const _Iter& _It) noexcept {
+    _STL_INTERNAL_STATIC_ASSERT(contiguous_iterator<_Iter>);
+    return reinterpret_cast<_Target*>(_STD to_address(_It));
+}
+#else // ^^^ defined(__cpp_lib_concepts) ^^^ / vvv !defined(__cpp_lib_concepts) vvv
+// When concepts aren't available, we can detect pointers. (Iterators should be unwrapped before using this.)
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>;
+
+template <class _Target, class _Iter>
+_NODISCARD constexpr _Target* _To_pointer(const _Iter& _It) noexcept {
+    _STL_INTERNAL_STATIC_ASSERT(is_pointer_v<_Iter>);
+    return reinterpret_cast<_Target*>(_It);
+}
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^
+
 template <class _InIt, class _OutIt>
 _CONSTEXPR20 _OutIt copy(_InIt _First, _InIt _Last, _OutIt _Dest) { // copy [_First, _Last) to [_Dest, ...)
     _Adl_verify_range(_First, _Last);
     const auto _UFirst = _Get_unwrapped(_First);
     const auto _ULast  = _Get_unwrapped(_Last);
     const auto _UDest  = _Get_unwrapped_n(_Dest, _Idl_distance<_InIt>(_UFirst, _ULast));
+
+#if _ITERATOR_DEBUG_LEVEL == 2
+    if constexpr (_Iterators_are_contiguous<_Unwrapped_t<_InIt>, _Unwrapped_t<_OutIt>>) {
+#ifdef __cpp_lib_is_constant_evaluated
+        if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated
+        {
+            _STL_VERIFY(
+                _UDest + (_ULast - _UFirst) <= _UFirst || _ULast <= _UDest, "ranges should not overlap each other");

```

Comment:
This condition and explanation string are incorrect. Somewhat unusually, `std::copy` permits overlapping in one direction. The Standardese is N4878 [alg.copy]/2 "*Preconditions:* `result` is not in the range `[first, last)`."

Forbidding the beginning of the output range from overlapping the input range prevents the initially copied element from overwriting input that needs to be read later. However, it is permitted for the output range to begin slightly before the input range, such that the end of the output range overlaps the beginning of the input range. This is allowed because the input range will be consumed before being overwritten.

(Note that these semantics differ from `memcpy` which forbids any overlapping and `memmove` which permits all overlapping.)

Additionally, this indicates that we're missing test coverage for this partial-overlapping scenario.

Reply:
This is perfectly correct. I fixed it and I added some tests to cover the copying partial-overlapping scenario. Thank you for the explanations.